File: day22_part1_2018.dart

import 'dart:io';

void main() {
  List<String> lines = File('input.txt').readAsLinesSync();
  int depth = int.parse(lines[0].split(' ')[1]);
  List<int> target = lines[1].split(' ')[1].split(',').map(int.parse).toList();

  int riskLevel = 0;
  List<List<int>> erosionLevels = List.generate(target[1] + 1, (i) => List<int>.filled(target[0] + 1, 0));

  for (int y = 0; y <= target[1]; y++) {
    for (int x = 0; x <= target[0]; x++) {
      int geologicIndex;
      if ((x == 0 && y == 0) || (x == target[0] && y == target[1])) {
        geologicIndex = 0;
      } else if (y == 0) {
        geologicIndex = x * 16807;
      } else if (x == 0) {
        geologicIndex = y * 48271;
      } else {
        geologicIndex = erosionLevels[y][x - 1] * erosionLevels[y - 1][x];
      }
      int erosionLevel = (geologicIndex + depth) % 20183;
      erosionLevels[y][x] = erosionLevel;
      riskLevel += erosionLevel % 3;
    }
  }

  print(riskLevel);
}

File: training_data/reflection-tuning/day22_part1_2018.py
from functools import lru_cache

def read_input():
    with open('input.txt', 'r') as f:
        depth = int(f.readline().split(': ')[1])
        target = tuple(map(int, f.readline().split(': ')[1].split(',')))
    return depth, target

@lru_cache(maxsize=None)
def geologic_index(x, y, target):
    if (x, y) in [(0, 0), target]:
        return 0
    if y == 0:
        return x * 16807
    if x == 0:
        return y * 48271
    return erosion_level(x-1, y, target) * erosion_level(x, y-1, target)

@lru_cache(maxsize=None)
def erosion_level(x, y, target):
    return (geologic_index(x, y, target) + depth) % 20183

def region_type(x, y, target):
    return erosion_level(x, y, target) % 3

def calculate_risk_level(target):
    risk_level = 0
    for y in range(target[1] + 1):
        for x in range(target[0] + 1):
            risk_level += region_type(x, y, target)
    return risk_level

depth, target = read_input()
total_risk = calculate_risk_level(target)
print(total_risk)
